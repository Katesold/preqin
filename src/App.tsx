import "./App.css";
import { Link, Outlet } from "react-router-dom";

function App() {
  return (
    <div className="App">
      <Outlet />
      <Link to={`investor/1`}>investor</Link>
    </div>
  );
}

export default App;
