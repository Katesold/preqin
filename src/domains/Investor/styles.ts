import styled from "styled-components";

export const InvestorContainer = styled.div`
  position: relative;
  display: inline-block;
  text-align: center;
  h1 {
    padding: 20px;
  }
`;
