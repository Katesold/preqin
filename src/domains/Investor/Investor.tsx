import { useEffect, useState } from "react";

import { AssetClasses } from "../../types";
import Dropdown from "../../components/Dropdown/Dropdown";
import { InvestorContainer } from "./styles";
import { StyledTable } from "../../components/Table/styles";
import Table from "../../components/Table/Table";
import { useLocation } from "react-router-dom";
import { useQuery } from "@tanstack/react-query";

export const Investor = () => {
  const options = Object.values(AssetClasses);
  const location = useLocation();
  const [selectedOption, setSelectedOption] = useState("Select an option");
  const [tableData, setTableData] = useState([]);

  const fetchCommitment = async (investorId = 1, selectedOption: string) => {
    const assetClass = selectedOption.slice(0, 2);
    const asset = Object.keys(AssetClasses).find((a) => a.includes(assetClass));

    const response = await fetch(
      `https://api.preqin.com/api/Investor/commitment/${
        asset ? asset.toLowerCase() : ""
      }/${investorId}`,
      {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${process.env.REACT_APP_TOKEN}`,
        },
        method: "GET",
      }
    );
    if (!response.ok) {
      throw new Error("Failed to fetch insvestor commitment");
    }

    return response.json();
  };

  const { data, isLoading, isError, error, isFetching } = useQuery({
    queryKey: ["commitment"],
    queryFn: () => fetchCommitment(location.state.firmID, selectedOption),
    enabled: !!location?.state?.firmID && selectedOption !== "Select an option",
  });

  const handleSelect = (option: string) => {
    setSelectedOption(option);
  };

  useEffect(() => {
    if (isError && error) {
      window.alert(error);
    }
  }, [isError, error]);

  useEffect(() => {
    if (selectedOption === "Select an option" || isError) {
      setTableData([]);
    } else {
      setTableData(data?.data);
    }
  }, [selectedOption, data?.data, isError]);

  return (
    <InvestorContainer>
      <h1>Investor {location?.state?.firmName}</h1>
      {location?.state ? <Table firms={[location?.state]} readOnly /> : null}
      <Dropdown
        options={options}
        onSelect={handleSelect}
        selectedOption={selectedOption}
      />
      {isLoading || isFetching ? (
        <p>Loading..</p>
      ) : (
        <div>
          <h2 style={{ padding: "20px" }}>Fund Details</h2>
          <StyledTable style={{ width: "50%", margin: "auto" }}>
            <tbody>
              {tableData?.length
                ? tableData.map(
                    (investor: { [s: string]: unknown } | ArrayLike<unknown>) =>
                      Object.entries(investor).map(([key, value]) => (
                        <tr key={key + value}>
                          <td
                            style={{
                              padding: "10px",
                            }}
                          >
                            {key}
                          </td>
                          <td>{value as string} </td>
                        </tr>
                      ))
                  )
                : null}
            </tbody>
          </StyledTable>
        </div>
      )}
    </InvestorContainer>
  );
};
