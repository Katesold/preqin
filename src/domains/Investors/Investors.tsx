import { memo, useEffect, useState } from "react";

import Table from "../../components/Table/Table";
import { StyledTableContainer } from "./styles";

export const Investors = memo(() => {
  const [firmsData, setFirmsData] = useState<{ data: any[] }>({ data: [] });

  const getData = async () => {
    try {
      const response = await fetch(
        "https://api.preqin.com/api/Investor?firmId=2670,2792,3320,3611",
        {
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${process.env.REACT_APP_TOKEN}`,
          },
          method: "GET",
        }
      );
      if (response.status === 200) {
        const result = await response.json();
        setFirmsData(result);
      }
    } catch (error) {
      window.alert("Error");
    }
  };

  useEffect(() => {
    getData();
  }, []);

  return (
    <StyledTableContainer>
      <h1>Investors List</h1>
      <Table firms={firmsData.data} />
    </StyledTableContainer>
  );
});
