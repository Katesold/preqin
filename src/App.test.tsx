import { render, screen } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router-dom";

import App from "./App";
import { Investor } from "./domains/Investor/Investor";
import { Investors } from "./domains/Investors/Investors";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
}));

test("renders App with table container", () => {
  const routes = [
    {
      path: "/",
      element: <App />,
      children: [
        {
          path: "investors/:Id",
          element: <Investor />,
        },
        {
          path: "investors",
          element: <Investors />,
        },
      ],
    },
  ];

  const router = createMemoryRouter(routes, {
    initialEntries: ["/investors"],
    initialIndex: 0,
  });

  const { container } = render(<RouterProvider router={router} />);
  expect(screen.getByText("Investors List")).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
