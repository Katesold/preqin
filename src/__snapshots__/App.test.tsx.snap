// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App with table container 1`] = `
<div>
  <div
    class="sc-eTNRI iBUEiE"
  >
    <h1>
      Investors List
    </h1>
    <table
      class="sc-iBdnpw bqszkd"
    >
      <thead
        class="sc-fsYfdN fQScNy"
      >
        <tr
          class="sc-kFCroH ebMFOg"
        >
          <th
            class="sc-jsEeTM clubGg"
          >
            Firm ID
          </th>
          <th
            class="sc-jsEeTM clubGg"
          >
            Firm Name
          </th>
          <th
            class="sc-jsEeTM clubGg"
          >
            Type
          </th>
          <th
            class="sc-jsEeTM clubGg"
          >
            Date Added
          </th>
          <th
            class="sc-jsEeTM clubGg"
          >
            Address
          </th>
        </tr>
      </thead>
      <tbody
        class="sc-qZrbh gUswMD"
      />
    </table>
  </div>
</div>
`;
